applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/d+d-constitutive_law_integrators/generic_compression_constitutive_law_integrator.h

```diff
@@ -0,0 +1,331 @@
+ * - SOFTENING_TYPE: The fosftening behaviour considered (linear, exponential,etc...)
+ * @tparam TYieldSurfaceType The yield surface considered
+ * @author Alejandro Cornejo 
+ */
+template <class TYieldSurfaceType>
+class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) GenericCompressionConstitutiveLawIntegratorDplusDminusDamage
+{
+  public:
+
+    ///@name Type Definitions
+    ///@{
+
+    /// The type of yield surface
+    typedef TYieldSurfaceType YieldSurfaceType;
+
+    /// The define the working dimension size, already defined in the yield surface
+    static constexpr SizeType Dimension = YieldSurfaceType::Dimension;
+
+    /// The define the Voigt size, already defined in the yield surface
+    static constexpr SizeType VoigtSize = YieldSurfaceType::VoigtSize;
+    
+    /// The type of plastic potential
+    typedef typename YieldSurfaceType::PlasticPotentialType PlasticPotentialType;
+
+    /// Counted pointer of GenericCompressionConstitutiveLawIntegratorDplusDminusDamage
+    KRATOS_CLASS_POINTER_DEFINITION(GenericCompressionConstitutiveLawIntegratorDplusDminusDamage);
+
+    /// Initialization constructor
+    GenericCompressionConstitutiveLawIntegratorDplusDminusDamage()
+    {
+    }
+
+    /// Copy constructor
+    GenericCompressionConstitutiveLawIntegratorDplusDminusDamage(GenericCompressionConstitutiveLawIntegratorDplusDminusDamage const &rOther)
+    {
+    }
+
+    /// Assignment operator
+    GenericCompressionConstitutiveLawIntegratorDplusDminusDamage &operator=(GenericCompressionConstitutiveLawIntegratorDplusDminusDamage const &rOther)
+    {
+        return *this;
+    }
+
+    /// Destructor
+    virtual ~GenericCompressionConstitutiveLawIntegratorDplusDminusDamage()
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method integrates the predictive stress vector with the CL using linear or exponential softening
+     * @param PredictiveStressVector The predictive stress vector
+     * @param UniaxialStress The equivalent uniaxial stress
+     * @param Damage The internal variable of the damage model
+     * @param Threshold The maximum uniaxial stress achieved previously
+     * @param rValues Parameters of the constitutive law
+     * @param CharacteristicLength The equivalent length of the FE
+     */
+    static void IntegrateStressVector(
+        array_1d<double, VoigtSize>& rPredictiveStressVector,
+        const double UniaxialStress,
+        double& rDamage,
+        double& rThreshold,
+        ConstitutiveLaw::Parameters& rValues,
+        const double CharacteristicLength
+        )
+    {
+        const Properties& r_material_properties = rValues.GetMaterialProperties();
+        const int softening_type = r_material_properties[SOFTENING_TYPE];
+        double damage_parameter;
+        TYieldSurfaceType::CalculateDamageParameter(rValues, damage_parameter, CharacteristicLength);

```

Comment:
Here is actually the caveat - the yield surface needs to know if it is in tension or compression for the selection of the right Fracture Energy.

Any ideas?
@AlejandroCornejo @loumalouomega

Reply:
The current solution is not very elegant, but it works :P

Reply:
Well, actually it does not really. For brittle materials (which is the main use of this material law) the fracture energy differes with factors around 100. Should we introduce a local Parameters here?

Reply:
I'll come up with some ideas and we will discuss!

Thank you